### PR TITLE
[GPU] Optimize out Gather by converting to implicit crop

### PR DESCRIPTION
**Pull request by byungilm**
+ Changed Gather if it cropped input along batch
+ Converted Gather to cldnn Crop in CreateGatherOpBase

### Details:
 - Converted Gather to cldnn Crop in CreateGatherOpBase in ops/gather
 - Optimized Crop as implicit crop

### Tickets:
 - 104425

### CODE REVIEW

File: src/plugins/intel_gpu/src/graph/graph_optimizer/prepare_buffer_fusing.cpp [hunk 1]
```diff
@@ -388,40 +455,17 @@ void prepare_buffer_fusing::run(program& p) {
...
-                        return;
-                    if (input_layout.data_padding.lower_size().batch[0] != 0 || input_layout.data_padding.upper_size().batch[0] != 0 ||
-                        input_layout.data_padding.lower_size().spatial[0] != 0 || input_layout.data_padding.upper_size().spatial[0] != 0 ||
-                        input_layout.data_padding.lower_size().spatial[1] != 0 || input_layout.data_padding.upper_size().spatial[1] != 0)
-                        return;
-                    // oneDNN doesn't support paddings
-                    if (usr->get_preferred_impl_type() == impl_types::onednn)
-                        return;
-                }
+                if (is_optimizable_padding_for_crop(node)) {
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/graph/graph_optimizer/prepare_buffer_fusing.cpp:462]: what about style below:
```
if (!is_optimizable_padding_for_crop)())
   return;
```

Other code uses such style and it keeps indent small.

File: src/plugins/intel_gpu/src/graph/graph_optimizer/prepare_buffer_fusing.cpp [hunk 2]
```diff
@@ -388,40 +455,17 @@ void prepare_buffer_fusing::run(program& p) {
...
+                    const auto& crop_layout = node.get_output_layout();
+                    const auto& crop_size = crop_layout.get_tensor();
+                    const auto& out_padd = crop_layout.data_padding;
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/graph/graph_optimizer/prepare_buffer_fusing.cpp:465]: nit: `padd` -> `pad`

**Review by isanghao — COMMENTED** [review]
looks good in general. let me check performance.

File: src/plugins/intel_gpu/src/graph/graph_optimizer/prepare_buffer_fusing.cpp (hunk 1 above)
Comment (byungilm) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/prepare_buffer_fusing.cpp:462]: Applied

File: src/plugins/intel_gpu/src/graph/graph_optimizer/prepare_buffer_fusing.cpp (hunk 2 above)
Comment (byungilm) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/prepare_buffer_fusing.cpp:465]: Modified. Also applied to old code.

File: src/plugins/intel_gpu/tests/unit/test_cases/gather_gpu_test.cpp [hunk 3]
```diff
@@ -2097,3 +2097,43 @@ TEST(gather_gpu_u8, 322_axisF) {
...
+    network.set_input_data("InputText", input2);
+
+    auto outputs = network.execute();
+
+    auto output = outputs.at("reorder").get_memory();
+    cldnn::mem_lock<int8_t> output_ptr(output, get_test_stream());
+
+    std::vector<int8_t> expected_results = {
+        5, 6, 7, 8
+    };
```
Comment (isanghao) [inline src/plugins/intel_gpu/tests/unit/test_cases/gather_gpu_test.cpp:2138]: what is the purpose of this test? It is just generated as gather, not crop..

File: src/plugins/intel_gpu/tests/unit/test_cases/crop_gpu_test.cpp [hunk 4]
```diff
@@ -1576,3 +1576,33 @@ TEST(crop_gpu, optimized_out_crop) {
...
+    network.set_input_data("Input", input1);
+
+    auto outputs = network.execute();
+
+    auto output = outputs.at("reorder").get_memory();
+    cldnn::mem_lock<int8_t> output_ptr(output, get_test_stream());
+
+    std::vector<int8_t> expected_results = {
+        5, 6, 7, 8
+    };
```
Comment (isanghao) [inline src/plugins/intel_gpu/tests/unit/test_cases/crop_gpu_test.cpp:1607]: could you add crop test for opt-out?

File: src/plugins/intel_gpu/tests/unit/test_cases/gather_gpu_test.cpp (hunk 3 above)
Comment (byungilm) [reply src/plugins/intel_gpu/tests/unit/test_cases/gather_gpu_test.cpp:2140]: As I explained shortly, I wanted to check processing of B axis cropping by both gather and crop.
I also added similar test into crop_gpu_test.
Expected results of can_be_optimized are different.

File: src/plugins/intel_gpu/tests/unit/test_cases/crop_gpu_test.cpp (hunk 4 above)
Comment (byungilm) [reply src/plugins/intel_gpu/tests/unit/test_cases/crop_gpu_test.cpp:1609]: This test also pass through the logic.
[ RUN      ] crop_single_axis.simple_Baxis
    -- prepare_buffer_fusing::run
    -- can_crop_be_optimized_along_batch()

Comment (byungilm) [reply src/plugins/intel_gpu/tests/unit/test_cases/crop_gpu_test.cpp:1609]: Added test into 'prepare_buffer_fusing'

**Review by isanghao — APPROVED** [review]

**Pull request merged**